utils/src/channels/ring.rs

```diff
@@ -0,0 +1,412 @@
+        // This channel never applies backpressure - it drops the oldest item instead.
+        Poll::Ready(Ok(()))
+    }
+
+    fn start_send(self: Pin<&mut Self>, item: T) -> Result<(), Self::Error> {
+        let mut shared = self.shared.lock().unwrap();
+
+        if shared.receiver_dropped {
+            return Err(SendError(item));
+        }
+
+        let old_item = if shared.buffer.len() >= shared.capacity {
+            shared.buffer.pop_front()
+        } else {
+            None
+        };
+
+        shared.buffer.push_back(item);
+        let waker = shared.receiver_waker.take();
+        drop(shared);
+
+        // Drop the old item after the lock is released to avoid potential mutex poisoning
+        drop(old_item);
+
+        if let Some(w) = waker {
+            w.wake();
+        }
+
+        Ok(())
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // No buffering in the sender - items are sent immediately to the shared buffer
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_close(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // Closing is handled by Drop
+        Poll::Ready(Ok(()))
+    }
+}
+
+/// The receiving half of a ring channel.
+///
+/// Implements [`Stream`] and [`FusedStream`] for receiving items. Use
+/// [`StreamExt::next`](futures::StreamExt::next) to receive items asynchronously.
+///
+/// The stream terminates (returns `None`) when all senders have been dropped
+/// and all buffered items have been consumed.
+pub struct Receiver<T> {
+    shared: Arc<Mutex<Shared<T>>>,
+}
+
+impl<T> Stream for Receiver<T> {
+    type Item = T;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut shared = self.shared.lock().unwrap();
+
+        if let Some(item) = shared.buffer.pop_front() {
+            return Poll::Ready(Some(item));
+        }
+
+        if shared.sender_count == 0 {
+            return Poll::Ready(None);
+        }
+
+        if !shared
+            .receiver_waker
+            .as_ref()
+            .is_some_and(|w| w.will_wake(cx.waker()))
+        {
+            shared.receiver_waker = Some(cx.waker().clone());
+        }
+        Poll::Pending
+    }
+}
+
+impl<T> FusedStream for Receiver<T> {
+    fn is_terminated(&self) -> bool {
+        let shared = self.shared.lock().unwrap();
+        shared.sender_count == 0 && shared.buffer.is_empty()
+    }
+}
+
+impl<T> Drop for Receiver<T> {
+    fn drop(&mut self) {
+        let Ok(mut shared) = self.shared.lock() else {
+            return;
+        };
+        shared.receiver_dropped = true;
+    }
+}
+
+/// Creates a new ring channel with the specified capacity.
+///
+/// Returns a ([`Sender`], [`Receiver`]) pair. The sender can be cloned to create
+/// multiple producers.
+///
+/// # Panics
+///
+/// Panics if `capacity` is zero.
```

Comment:
Use `NZUsize`? (I suppose you did this for compatibility with how other channels are typically initialized?)